lib/core/src/Cardano/Wallet.hs

```diff
@@ -1306,50 +1267,105 @@ selectAssets
         , HasNetworkLayer ctx
         )
     => ctx
-    -> WalletId
+    -> (WalletId, UTxOIndex, Wallet s, Set Tx)
     -> TransactionCtx
-    -> NonEmpty TxOut
+    -> [TxOut]
     -> (s -> SelectionResult TokenBundle -> result)
     -> ExceptT ErrSelectAssets IO result
-selectAssets ctx wid tx outs transform = do
-    (cp, _, pending) <- withExceptT ErrSelectAssetsNoSuchWallet $
-        readWallet @ctx @s @k ctx wid
-    let s = getState cp
-
-    guardWithdrawal pending
-
-    pp <- liftIO $ currentProtocolParameters nl
+selectAssets ctx' w' tx' outputs' transform' =
+    case outputs' of
+        [] -> selectAssetsNoOutputs ctx' w' tx' transform'
+        x:xs -> selectAssetsWithOutputs ctx' w' tx' (x :| xs) transform'
 
-    let utxo :: UTxOIndex
-        utxo = UTxOIndex.fromUTxO $ availableUTxO @s pending cp
-
-    liftIO $ traceWith tr $ MsgSelectionStart utxo outs
-    sel <- performSelection
-        (calcMinimumCoinValue tl pp)
-        (calcMinimumCost tl pp tx)
-        (initSelectionCriteria tl pp tx utxo outs)
-    liftIO $ traceWith tr $ MsgSelectionDone sel
-    withExceptT ErrSelectAssetsSelectionError $ except (transform s <$> sel)
   where
-    nl = ctx ^. networkLayer
-    tl = ctx ^. transactionLayer @k
-    tr = ctx ^. logger
-
-    -- Ensure that there's no existing pending withdrawals. Indeed, a withdrawal
-    -- is necessarily withdrawing rewards in their totality. So, after a first
-    -- withdrawal is executed, the reward pot is empty. So, to prevent two
-    -- transactions with withdrawals to go through (which will inevitably cause
-    -- one of them to never be inserted), we warn users early on about it.
-    guardWithdrawal :: Set Tx -> ExceptT ErrSelectAssets IO ()
-    guardWithdrawal pending = do
-        case Set.lookupMin $ Set.filter hasWithdrawal pending of
-            Just pendingWithdrawal | withdrawalToCoin (txWithdrawal tx) /= Coin 0 ->
-                throwE $ ErrSelectAssetsAlreadyWithdrawing pendingWithdrawal
-            _otherwise ->
-                pure ()
+    selectAssetsNoOutputs
+        :: ctx
+        -> (WalletId, UTxOIndex, Wallet s, Set Tx)
+        -> TransactionCtx
+        -> (s -> SelectionResult TokenBundle -> result)
+        -> ExceptT ErrSelectAssets IO result
+    selectAssetsNoOutputs ctx w@(wid, _, _, _)tx transform = do
```

Comment:
Arg. Maybe I should have picked a better name in the first place because, the fact that there are no outputs is not actually what drives the body of this function. That selection really is selecting with delegation / key de-registration in mind. So it is more like `selectAssetsForDelegation`... Conflating this with the `selectAssets` function itself feels wrong and confusing :/ 

What exactly is the rationale for merging these?

Reply:
> Unify selectAssets and selectAssetsNoOutputs
> Idea was to allow estimateFee to call both readUTxOIndex and selectAssets — simplifying the call-sites — but I didn't end up doing this now. This seems right to me still, unless I have missed something?

Reply:
I don't think the call site would be much more complicated by keeping them separate, no ?

Reply:
Right now we do:
```
wal <- unsafeRunExceptT $ W.readUTxOIndex @_ @s @k w wid
let runSelection = W.selectAssets @_ @s @k w wal txCtx (out :| []) getFee
runExceptT $ withExceptT show $ W.estimateFee runSelection
```

But I was thinking we could get to:
```
runExceptT $ withExceptT show $ W.estimateFee @_ @s @k w txCtx [out] getFee
```
With `NoOutputs` and `Outputs` represented by the list argument, we don't need to pass in the right variant of `selectAssets`. `estimateFee` could then call `readUTxOIndex` directly (but outside the loop).

Reply:
But choosing the right variant of `selectAssets` is actually quite crucial because, it is the caller who knows what the right variant is. So this:

```
wal <- unsafeRunExceptT $ W.readUTxOIndex @_ @s @k w wid
let runSelection = W.selectAssets @_ @s @k w wal txCtx (out :| []) getFee
runExceptT $ withExceptT show $ W.estimateFee runSelection
```

Is much better.

Reply:
(modulo the `unsafeRunExceptT` :grimacing: ...)